### DISCUSSION TITLE
Setup for vim-slime with lazy.nvim for REPL interaction

### DISCUSSION THREAD

**Discussion by sourproton** [LazyVim/LazyVim#234] [Show and tell]
EDIT: I now use this simple plugin I wrote https://github.com/sourproton/tunnell.nvim

I have a Python and Julia REPL based workflow and I achieved a pretty nice setup with [LazyVim](https://github.com/LazyVim/LazyVim), [vim-slime](https://github.com/jpalardy/vim-slime) and [tmux](https://github.com/tmux/tmux/wiki).

In my `~/.config/nvim/lua/plugins/coding.lua` file I have
```lua
return {
    -- slime (REPL integration)
    {
        "jpalardy/vim-slime",
        keys = {
            { "<leader>rc", "<cmd>SlimeConfig<cr>",               desc = "Slime Config" },
            { "<leader>rr", "<Plug>SlimeSendCell<BAR>/^# %%<CR>", desc = "Slime Send Cell" },
        },
        config = function()
            vim.g.slime_target = "tmux"
            vim.g.slime_cell_delimiter = "# %%"
            vim.g.slime_bracketed_paste = 1
        end,
    },
}
```
When coding, I divide my cells with comments starting with `# %%`, like this
```python
# %% imports
from foobar import bar
from foobaz import baz

# %% helpers
def spam(a, b):
    """Does nothing useful"""
    c = bar(a) + baz(b)

    return c**2

# %% calculations
x = spam(0, 1)

# %% displaying results
print(x)
```
I also have the code in a tmux pane, and a REPL in another pane on the same window. To start the REPL interaction, I run the `<leader>rc` shortcut which triggers SlimeConfig, and usually set the `tmux socket` to default. If the code pane is in the left and the REPL is in the right, then the `tmux target pane` should be `:.1`. More on that on the [vim-slime repo readme](https://github.com/jpalardy/vim-slime/blob/main/README.md#tmux).

Then, to send the cells to the REPL, I run the `<leader>rr` shortcut. The way it is setup makes so that it automatically searches and jumps to the next `# %%` cell start, via the vim search, which means that `n` and `N` also jump between cells so it's easy to modify and run the cell again, or skip one.

Finally, `vim.g.slime_bracketed_paste = 1` makes it possible to code with blank lines (like in the `spam` function definition and not break the REPL (works great with Python, IPython and Julia, didn't test others).

Anyways here's my workflow, I hope you find it useful and please share your solutions to achieve your REPL based workflows!

**Comment by gaardhus** [comment]
Great stuff!

I just modified the `rc` to `rC` since it was used by another keymap by default. Also added the possability to send specific lines in visual mode:


```lua
return {
  -- slime (REPL integration)
  {
    "jpalardy/vim-slime",
    keys = {
      { "<leader>rC", "<cmd>SlimeConfig<cr>", desc = "Slime Config" },
      { "<leader>rr", "<Plug>SlimeSendCell<BAR>/^# %%<CR>", desc = "Slime Send Cell" },
      { "<leader>rr", ":<C-u>'<,'>SlimeSend<CR>", mode = "v", desc = "Slime Send Selection" },
    },
    config = function()
      vim.g.slime_target = "tmux"
      vim.g.slime_cell_delimiter = "# %%"
      vim.g.slime_bracketed_paste = 1
    end,
  },
}
```

**Reply by sourproton** [reply]
Very nice!

Since then, I made this plugin, which you might find helpful:

https://github.com/sourproton/tunnell.nvim
